chalk-ir/src/visit/boring_impls.rs

```diff
@@ -0,0 +1,341 @@
+//! This module contains "rote and uninteresting" impls of `Visit` for
+//! various types. In general, we prefer to derive `Visit`, but
+//! sometimes that doesn't work for whatever reason.
+//!
+//! The more interesting impls of `Visit` remain in the `visit` module.
+
+use crate::{
+    AssocTypeId, DebruijnIndex, Goals, ImplId, Interner, Parameter, ParameterKind,
+    PlaceholderIndex, ProgramClause, QuantifierKind, StructId, Substitution, SuperVisit, TraitId,
+    UniverseIndex, Visit, VisitResult, Visitor,
+};
+use chalk_engine::{context::Context, ExClause, FlounderedSubgoal, Literal};
+use std::{marker::PhantomData, sync::Arc};
+
+impl<T: Visit<I>, I: Interner> Visit<I> for &T {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        (**self).visit_with(visitor, outer_binder)
+    }
+}
+
+impl<T: Visit<I>, I: Interner> Visit<I> for Vec<T> {
+    fn visit_with<'i, R: VisitResult>(
+        &self,
+        visitor: &mut dyn Visitor<'i, I, Result = R>,
+        outer_binder: DebruijnIndex,
+    ) -> R
+    where
+        I: 'i,
+    {
+        let mut result = R::new();
+        for e in self {
+            result = result.and_then(|| e.visit_with(visitor, outer_binder))
```

Comment:
Hmm, so, the one problem here is that we are not giving the `and_then` the ability to stop iterating through the vector. I think we'd have to write this to "unfold" across the stack like so:

```rust
fn unroll<I: Interner, T: Visit<I>, R: VisitResult>(
    visitor: &mut dyn Visitor<'_, I, Result = R>,
    iter: impl Iterator<Item = T>
) -> R {
    let mut iterator = self.iter();
    match iterator.next() {
        None => R::new(),
        Some(elem) => elem.visit_with(visitor, outer_binder).and_then(|| unroll(visitor, iter))
    }
})
```

But seeing how nasty this is makes me wonder if `and_then` is *really* the right primitive, versus something that returns e.g. a `Result`-like enum that means "keep going", or perhaps adding a boolean to the result like `return_early`, so that the loop would be like:

```rust
for e in self {
    result = result.combine(e.visit_with(visitor, outer_binder));
    if result.return_early() { return result; }
}
result
```

Reply:
I thought about this when implementing, and I chose to do it that way instead of unrolling because it matches the behavior of the regular struct Visit.
```
fn visit_with(...) -> R {
  R::new()
      .and_then(|| a.visit()
                            .and_then(|| b.visit()
                                                  .and_then(|| c.visit())))
}
```
would be the struct visit rewritten as unrolling, which is kind of a weird mix of right and left folding.

Replacing `and_then` with `combine` + boolean sounds very annoying, since you would need to call it after each visit of a struct member too. I would just add an `return_early` method as a perf hack for iterator things, which is a little inconvenient since it would require `and_then` and `return_early` behavior to match (but when implementing `VisitResult` you will probably use `return_early` in `and_then` anyway, so who cares). Alternatively, we can hack the existing implementation by doing 
```
for e in self {
    let visited = false;
    result = result.and_then(|| { e.visit_with(visitor, outer_binder); visited = true; })
    if !visited {
       break
    }
}
```
If we propose a contract that `and_then` not calling the closure means early return, then it kinda feels less hacky.

Reply:
I feel like I'd prefer to have one mechanism, let's call it `return_early`. If we want, we can build the `and_then` mechanism on top of that readily enough:

```rust
fn and_then(self, op: impl FnOnce() -> Self) -> Self {
    if self.return_early() { self } else { self.combine(op()) }
}
```

That said, I'd probably change the deriving at least to use `return_early` instead of `and_then` because it'll be more efficient at runtime and requires less compiler tricks to compile.